### DISCUSSION TITLE
Is accessing the db on the client-side something that should even be considered in a real-world scenario?

### DISCUSSION THREAD

**Discussion by caliguian** [supabase/supabase#29730] [Questions]
As some background, I come with 20 years of db experience but I have never used a db that can be accessed through javascript before now. I realize that I am likely just missing something important, but in all of my many hours of reading the docs, watching the videos, and creating test sites, I can't seem to figure out how to use this type of database in any sort of secure way. Maybe I just missed seeing the correct documents or videos, and, if that's the case, I would love it if someone could provide links to those items for me.

Anyway... when you have a db where everything is handled on the backend, security is pretty straightforward; but, when everything is accessible by anyone through a js interface it seems to come with a lot of gotchas, and it kinda feels like everything that's in place for security is just a hack that leads to additional hacks being required. At first, I thought the RLS policies would probably handle most of the security, but I am struggling with the limitations of just using RLS policies.  My main concern is that if I push a project into production, some nefarious user will find ways to exploit the security vulnerabilities that inherently come with allowing db access through front-end code.

As a very simple example, I have seen multiple threads related to people being able to scrape publicly accessible tables, and that seems like a really big problem that Supabase can't really resolve. I saw the following attempt at resolving the issue, but it really doesn't seem to do much for the problem: https://github.com/orgs/supabase/discussions/28858

Another issue is that the RLS policies restrict people to specific ROWS, but not specific COLUMNS. So, to combat this, you can create Views, but since views use the RLS policies from the tables they are referencing, you still have to give the users access to those tables, which can then be accessed through the frontend as well. Or, if you create public tables with the columns you want generally accessible, and then create separate linked tables in non-public schemas with the private data, you now have twice as many tables, with twice as many RLS policies, and it all feels fairly convoluted. The, when something needs to be updated, you can no longer just do a simple update query, but instead you need to use custom functions and/or triggers since some of the data belongs in the public tables and some on the non-public tables. It seems to get pretty hairy, pretty quickly.

Like I said at the beginning, I'm guessing that I may just be missing something that makes it all make sense, but right now it feels fairly unusable in a real-world scenario. At least without making the db portion super-complicated and very much like spaghetti code, with 10 different custom things needed for almost every table in order to make it all work.

**Comment by GaryAustin1** [comment]
You can do either way with Supabase.  Lock it all down and do all from a server with the REST API or even a server using SQL directly with a Postgres library.  You can run an app with no server at all and use the REST API with RLS to protect the data.  Or anything in between.  

Certainly if you allow REST API access from a browser/app you have to make sure your RLS is sound for security.

You can also use rpc calls only to the database or Edge functions only to the database.

Like Firebase, Supabase initially was focused on serverless applications and both rely on JWT's and security policies to the data with no server in between.   But by blocking the API with various options you can do full server only access also.
